gateway/app/main/events/fl_events.py

```diff
@@ -0,0 +1,154 @@
+    """
+    data = message[MSG_FIELD.DATA]
+    response = {}
+
+    try:
+        # Retrieve JSON values
+        serialized_model = data.get(MSG_FIELD.MODEL, None)  # Only one
+        serialized_client_plans = data.get(CYCLE.PLANS, None)  # 1 or *
+        serialized_client_protocols = data.get(CYCLE.PROTOCOLS, None)  # 0 or *
+        serialized_avg_plan = data.get(CYCLE.AVG_PLAN, None)  # Only one
+        client_config = data.get(CYCLE.CLIENT_CONFIG, None)  # Only one
+        server_config = data.get(CYCLE.SERVER_CONFIG, None)  # Only one
+
+        # Create a new FL Process
+        fl_process = processes.create_process(
+            model=serialized_model,
+            client_plans=serialized_client_plans,
+            client_protocols=serialized_client_protocols,
+            server_averaging_plan=serialized_avg_plan,
+            client_config=client_config,
+            server_config=server_config,
+        )
+
+    except Exception as e:  # Retrieve exception messages such as missing JSON fields.
+        response[RESPONSE_MSG.ERROR] = str(e)
+
+    return json.dumps(response)
+
+
+def authenticate(message: dict, socket) -> str:
+    """ New workers should receive a unique worker ID after authenticate on PyGrid platform.
+        
+        Args:
+            message : Message body sended by some client.
+            socket: Socket descriptor.
+        Returns:
+            response : String response to the client
+    """
+    response = {}
+
+    # Create a new worker instance and bind it with the socket connection.
+    try:
+        # Create new worker id
+        worker_id = str(uuid.uuid4())
+
+        # Create a link between worker id and socket descriptor
+        handler.new_connection(worker_id, socket)
+
+        # TODO:
+        # Create a new User/Worker instance to map and manage this worker.
+
+        response[MSG_FIELD.WORKER_ID] = worker_id
+
+    except Exception as e:  # Retrieve exception messages such as missing JSON fields.
+        response[RESPONSE_MSG.ERROR] = str(e)
+
+    return json.dumps(response)
+
+
+def cycle_request(message: dict, socket) -> str:
+    """This event is where the worker is attempting to join an active federated learning cycle. 
+        
+        Args:
+            message : Message body sended by some client.
+            socket: Socket descriptor.
+        Returns:
+            response : String response to the client
+    """
+    data = message[MSG_FIELD.DATA]
+    response = {}
+
+    try:
+        # Retrieve JSON values
+        worker_id = data.get(MSG_FIELD.WORKER_ID, None)
+        model_id = data.get(MSG_FIELD.MODEL, None)
+        version = data.get(CYCLE.VERSION, None)
+        ping = data.get(CYCLE.PING, None)
+        download = data.get(CYCLE.DOWNLOAD, None)
+        upload = data.get(CYCLE.UPLOAD, None)
+
+        # TODO:
+        # Search/Identify worker id
+        # Search/Identify model id
+        # Check worker health (ping/download/upload rate)
```

Comment:
I pretty liked your idea!
Thinking about it, I tried to create an ephemeral structure to manage the cycles.

This structure should 
   - **manage workers registered in this cycle**,
   - **generate and manage ephemeral hash codes to download models / protocols**
   - **control the cycle life time**.

I believe that it is not necessary to save this structure in a database. Since it will only be instantiated during the execution of a cycle(should be deleted after cycle ends), and we will already be saving the training configs by storing the FLProcess.

What do you think about it?  @cereallarceny , @Prtfw,  @hericlesme

Reply:
I definitely want to have the philosophy of "everything that can be saved in a database should be stored in a database". This will allow us to build a UI in the future to control the federated learning process and show administrators of a PyGrid gateway "how many workers apply", "how many workers are accepted but don't report (because of drop off or something)", "how what percentage of workers are declined", etc. This are incredibly helpful metrics and statistics for a model developer to see and would allow them the opportunity to tweak their `server_config` to get the best results. If they don't have data on this... they'll never know WHY their model isn't being trained properly. :)

Any statistic you have - store it.